weave/flow/eval2.py

```diff
@@ -0,0 +1,200 @@
+from collections.abc import Iterator
+from contextlib import contextmanager
+from contextvars import ContextVar
+from types import MethodType
+from typing import Any, cast
+
+from pydantic import BaseModel, ConfigDict, PrivateAttr
+
+import weave
+from weave.flow.eval import Evaluation, default_evaluation_display_name
+from weave.flow.model import Model
+from weave.flow.scorer import Scorer
+from weave.trace.context import call_context
+from weave.trace.context.weave_client_context import require_weave_client
+from weave.trace.weave_client import Call
+
+ID = str
+
+NOT_DEFINED = "Not defined for custom scoring"
+
+# Context variable to store the current output safely between threads.  This also
+# ensures that only 1 version of the predict method is saved because the code
+# contents are always the same.
+current_output: ContextVar[Any] = ContextVar("current_output", default=None)
+current_score: ContextVar[float | None] = ContextVar("current_score", default=None)
+current_summary: ContextVar[dict | None] = ContextVar("current_summary", default=None)
+
+
+@contextmanager
+def set_current_output(output: Any) -> Iterator[None]:
+    """Set the current output in a thread-safe way using context variables."""
+    token = current_output.set(output)
+    try:
+        yield
+    finally:
+        current_output.reset(token)
+
+
+@contextmanager
+def set_current_score(score: float) -> Iterator[None]:
+    token = current_score.set(score)
+    try:
+        yield
+    finally:
+        current_score.reset(token)
+
+
+@contextmanager
+def set_current_summary(summary: dict) -> Iterator[None]:
+    token = current_summary.set(summary)
+    try:
+        yield
+    finally:
+        current_summary.reset(token)
+
+
+class BetaScoreLogger(BaseModel):
+    prediction_id: ID
+    score: float
+    metadata: dict | None
+
+
+class BetaPredictionLogger(BaseModel):
+    model_id: ID
+    inputs: dict
+    output: Any
+    hack_reference_to_predict_and_score: Call
+    hack_reference_to_evaluate: Call
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    async def log_score(
+        self, scorer_name: str, score: float, metadata: dict | None = None
```

Comment:
similarly, i think scorer_name should support dictionaries

Reply:
I don't understand this one either, can you expand?